### ISSUE TITLE
react-fast-refresh is always added to the client

### ISSUE THREAD

**Issue by jankapunkt** [meteor/meteor#12086]
The package `react-fast-refresh` is always added to the client bundle, adding 32kb of unused code (unminified), even if `hot-module-replacement` is not present.

When I'm not using React then I don't need it.

Using `DISABLE_REACT_FAST_REFRESH` has no effect, the package is still added to the client.

Repro:

- create new Meteor project that not uses React
- run using `meteor`
- open chrome dev tools --> source --> coverage
- see react-fast-refresh being present but being marked as nearly 100% of unused code

My proposal - only add it when the project is created with React (making it an explicit package to be added and add it to the default `create` boilerplate) or making `DISABLE_REACT_FAST_REFRESH` having the effect, that `react-fast-refresh` is completely stripped of the bundling step.

**Comment by denihs** [comment]
I'd go with this approach `only add it when the project is created with React`.

**Comment by A-5ingh** [comment]
I can take this issue up if you can provide some more details about to the issue as I'm new to meteor. Thanks

**Comment by vaishnav-197** [comment]
Hey 
I Can take this up 👍

**Comment by Grubba27** [comment]
Sure! you can fill in a PR

**Comment by Rimpyyadav** [comment]
Anyone working on this issue?

**Comment by StorytellerCZ** [comment]
@Rimpyyadav feel free to work on it.

**Comment by Rimpyyadav** [comment]
I am using node18, for setting up this do i need to use node 14?

**Comment by jankapunkt** [comment]
Depending on the target release 2.x or 3.x this may target different node versions. From my end this should already be reflected in 2.x releases but I don't know if that's feasible @denihs 

However I think the development/contribution guide covers how to set up everything.

**Comment by Rimpyyadav** [comment]
@jankapunkt Thank you for the response!!

**Comment by denihs** [comment]
I'm going to work on this with @Grubba27

**Comment by jankapunkt** [comment]
@denihs I can also try to fix it and make a PR, so you guys can focus on 3.0 :-)

**Comment by jankapunkt** [comment]
I investogated the problem a bit. So the biggest issue is right now, that we simply can't have a dedicated react compiler as JSX can be written within .js or .mjs as well and there is no way to have multiple compiler plugins being registered for the same file types.

I furthermore checked and it's nearly impossible to unwire react-fast-refresh as it provides the babel plugin config, which is determined at project build time, so I can't make react-fast-refresh simply a weak dependency as it's hard-wired to the registered compiler.

From what I see there are some potential solutions:

Two packages
- remove all react-specific code from `ecmascript`, making it handle only .js and .mjs and have a dedicated `ecmascript-react`
- advantage: clean separation between react and non react file handling
- disadvantage: two ecmascript packages have to be maintained
- existing projects would have to replace `ecmascript` with `ecmascript-react`

Allow overrides
- Allow compiler plugins to override other compiler plugins, so we can have ecmascript as base and define a react-compiler that overrides file-handling for js and mjs and adds additional jsx
- advantage: no two packages that contain mostly the same code
- disadvantage: how to determine override order; could risk the whole arch to become flaky

Strict JSX
- remove jsx from ecmascript default copmiler and create an additional package `react-babel-compiler` that only adds .jsx files
- advantage: very clean separation, smallest changes
- disadvantage: files with jsx will have to strictly end with jsx

At this point I even did not consider .tsx files and typescript. That has to be investigated, too.

**Comment by nachocodoner** [comment]
This is a great opportunity to lean the Meteor client bundle for the next optimization releases.

When optimizing the bundle size for the next work with the bundler, I’ll consider this unless another contributor completes it first.

**Edit:** I’m not sure how this affects the modern stack using SWC . Will react-fast-refresh still work if it depends on Babel? HMR seems fine with SWC, but react-fast-refresh might miss something.

With [Meteor-RSPack using a modern bundler](https://forums.meteor.com/t/meteor-rspack-integration-a-modern-bundler-meets-meteor-3-4/63696), we could remove react-fast-refresh entirely at the Meteor level for these projects.

**Comment by nachocodoner** [comment]
Maybe if I achieve to implement a devLike option for packages, we can skip the react-fast-refresh and others from production final bundle, https://github.com/meteor/meteor/issues/10902#issuecomment-2984701425

**Comment by jankapunkt** [comment]
Still relevant until #13797 is merged

**Comment by jankapunkt** [comment]
Still relevant until #13797 is merge

**Comment by nachocodoner** [comment]
Meteor 3.4 is already out. When Rspack is enabled, `react-fast-refresh` is handled by Rspack for all React app code. This should reduce the need of the extra `react-fast-refresh` on other projects. I still need to double check if it can be safely removed for other project types.

As part of Meteor 3.4.1, I will verify if this is still relevant and if it adds noticeable overhead. I will close the issue or further tweak it for next release based on the results.

**Comment by jirkadev** [comment]
Yes, please make this package optional for codebases without React. It could save a few KBs from the initial bundle.
